### ISSUE TITLE
problem: no way to suspend users who post spam etc

### ISSUE THREAD

**Issue by gazhayes** [Blockrazor/blockrazor#800]
Problem: there will be some users (maybe even bots) that post spam or silly things. There will also be users who repeatedly submit bad bounties (e.g. adding a new coin and typing in random stuff in fields they can't be bothered to check), as well as lazy-answering in the ELO rankings.

possible solution:

When users flag a comment/feature/red flag/summary/whatever and the ratio between flags and likes reaches a certain point, that item should placed in the moderation queue. If 2 moderators uphold the flag status, the comment should be removed. The user who posted that item should then receive a strike on their profile.

The same thing should happen for bounties that aren't approved (bad wallet images, bad coin additions, etc). As well as when lazy answering is detected.

When a users has (these numbers are just random examples) 3 strikes in a week OR 6 strikes in a month, they should be suspended.

When an account is suspended, the user should be presented with a list of their infractions and given a chance to write a message asking for account reinstatement. The infractions along with the message should be presented to moderators, who can vote on whether or not to reinstate the account (I guess using the same algorithm as voting on pending changes to coin attributes: +3 and its approved, -3 and it's rejected. Total = upvotes - downvotes.

ping @anbud

**Comment by AshHimself** [comment]
To prevent bots (not full prevention) we could implement https://www.google.com/recaptcha.

All your other comments are good, it just should be coded in a scalable manner so we can tweak it as the site progresses if need be.

**Comment by gazhayes** [comment]
>  it just should be coded in a scalable manner so we can tweak it as the site progresses if need be.

Definitely yes.

Recapture is not a bad idea. When/where do you think it should be used?

**Comment by AshHimself** [comment]
You could just ask for it before commenting (once) to verify they are human or any other feature where we want to prevent user/bot automation from spamming the site.

**Comment by anbud** [comment]
@gazhayes, I agree, the proposed solution is quite sufficient for now. I also agree with @AshHimself, (re)captcha should be implemented as well, it's the basis of bot and spam prevention.

I can take this one if it's available.

**Comment by AshHimself** [comment]
Ok with me, i'm active on a few other tickets at the moment.

**Comment by gazhayes** [comment]
@anbud where/when do you think users should be prompted for captcha?

**Comment by anbud** [comment]
@gazhayes, comments are the most sensitive part I guess, as there's currently no limit on how many comments users can post. So, basically, we can ask them to verify once before posting comments, or even features/summaries/red flags. I'm not sure about other parts of the site, as most other stuff already has some kind of validation or verification.
My idea is to only ask them once of course, when they first try to post, and save it for the rest of the session to avoid bothering them.

**Comment by AshHimself** [comment]
From experience, it should only ask once, but if we suspect spamming it
should ask again. e.g. if you post 3 comments in 10 seconds it'll prompt
for another recaptcha.

**Comment by gazhayes** [comment]
We probably need to prevent users from posting more than once a week or something anyway for features/red flags/summaries/etc (wherever there's a reward for having the top comment, since new stuff immediately goes to the top). 

I'm almost certain I logged an issue about this very recently.

This however won't prevent cross-currency spamming. We could probably just store the timestamps of the last x activities (site wide) in the userdata collection for users and check how far apart they are with every new activity, and if they are too close prompt for a capture.

**Comment by anbud** [comment]
@gazhayes, I actually solved that issue with limiting yesterday. :)
You're right, we can save a few last activities and prompt the captcha screen if anything seems too fast.

**Comment by gazhayes** [comment]
>  I actually solved that issue with limiting yesterday. :)

Ha! I swear I have just a little bit more grey hair since the start of this project, just from trying to keep up.

**Comment by AshHimself** [comment]
haha.. its hard to keep up like.. I swear you don't sleep.

**Comment by gazhayes** [comment]
> captcha UI could use a little bit of love, and captcha could be possibly placed in a modal.

That probably deserves its own issue, I'll leave that to you to create once you're confident the rest of recapture is working and it's really only UI stuff.

**Comment by anbud** [comment]
Ok. I'll create one after #820 is merged, tested it on localhost now and it works fine, but I want to be sure it's working in production as well.

**Comment by gazhayes** [comment]
It's probably worth doing some tests for this too.

**Comment by anbud** [comment]
It's done, it's working. I'll create an issue right away.

I agree with testing, but this is client side, and client-side testing isn't really supported by Mocha.
We could possibly use something like [webdriver.io](http://webdriver.io/) for client-side (acceptance) tests.
